### ISSUE TITLE
QueryParamTypes fails type checking without "as const"

### ISSUE THREAD

**Issue by nickkelley42** [typed-ember/ember-cli-typescript#1295]
# Which package(s) does this problem pertain to?
  - [x] @types/ember__controller

### What are instructions we can follow to reproduce the issue?

In a controller, specifying a type on a query parameter fails type checking; for example:

```
export default class Roles extends Controller.extend({
}) {
  queryParams = [
    { something: { type: 'array' } },
  ];
}
```

##### Reproduction Case

**Link:** https://codesandbox.io/s/my-app-forked-n8nqj?file=/app/controllers/example.ts

### Now about that bug. What did you expect to see?

Expected: this should pass type checking

### What happened instead?

Fails type checking with the following message:

```
app/controllers/example.ts:7:3 - error TS2416: Property 'queryParams' in type 'Example' is not assignable to the same property in base type 'EmberClassArguments & Controller'.
  Type '{ something: { type: string; }; }[]' is not assignable to type '(string | Record<string, QueryParamConfig | undefined>)[]'.
    Type '{ something: { type: string; }; }' is not assignable to type 'string | Record<string, QueryParamConfig | undefined>'.
      Type '{ something: { type: string; }; }' is not assignable to type 'Record<string, QueryParamConfig | undefined>'.
        Property 'something' is incompatible with index signature.
          Type '{ type: string; }' is not assignable to type 'QueryParamConfig'.
            Types of property 'type' are incompatible.
              Type 'string' is not assignable to type '"string" | "number" | "boolean" | "array" | undefined'.

7   queryParams = [{ something: { type: 'array' }}];
```

The type in question is defined in `node_modules/@types/ember__controller/index.d.ts` as:
```
type QueryParamTypes = 'boolean' | 'number' | 'array' | 'string';
// ...
interface QueryParamConfig {
    type?: QueryParamTypes;
    scope?: QueryParamScopeTypes;
    as?: string;
}
// ...
queryParams: Array<string | Record<string, QueryParamConfig | undefined>>;
```

Adding `as const` to the type string makes it pass type checking:
```
export default class Example extends Controller.extend({
  // anything which *must* be merged to prototype here
}) {
  // normal class body definition here
  queryParams = [{ something: { type: 'array' as const }}];
}
```

**Comment by chriskrycho** [comment]
Thanks for writing this up! I don't think there's a good fix here other than adding `as const`: this is just a limitation of the compiler as it stands. However, I'm making a point to document it as I'm working on (finally!) finishing our post-Octane docs update.

**Comment by nickkelley42** [comment]
Awesome, thank you for taking the time to document this! This addon has been a huge help.

**Comment by MrChocolatine** [comment]
I just want to leave a trace, and track that issue, as I came across the same problem while playing with query parameters, I could not get the properties `type` and `scope` working.

But the workaround `as const` works indeed.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9bc42ea5d45eeafdc62fa2964cf76e0bafc9ca9e/types/ember__controller/index.d.ts#L16-L35

**Comment by MrChocolatine** [comment]
> @chriskrycho Chris Krycho closed this in 25bc558 2 days ago

Awesome for https://github.com/typed-ember/ember-cli-typescript/pull/935 ! Thanks.
